### PR TITLE
Simplify ZStream#flatMap using ZManaged.switchable

### PR DESCRIPTION
**Pull request by iravid**
And adding some missing ZManaged combinators.

### CODE REVIEW

**Comment by iravid** [comment]
cc @vasilmkd who notes the similarity of ZManaged.switchable to flatMap :-)

**Comment by mschuwalow** [comment]
Wow. This looks so much better :fireworks: 

One random thought: Couldn't we use the same idea in order to implement flatMapPar? We could keep a set of open streams around and pull from them in a round robin fashion / race them and keep extra results in a fixed size buffer. Once a stream is done we kick it out and open the next one.
This would also no longer require any queues :)
Old behavior with buffering of output could be easily restored using buffer.

I don't think switchable as is powerful enough to this, but I don't see why switchableN shouldn't be possible

**Comment by iravid** [comment]
@mschuwalow 
Interesting that you bring this up :-) I looked into this when we switched over to `Pull`. My thought then is that racing was less efficient than forking off the streams and having them push into a queue. So this is definitely possible and possibly simpler; worth exploring if you have the time!

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -1324,13 +1324,15 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     self.fork >>= (_.toFutureWith(f))
 
   /**
-   * Converts this ZIO to [[zio.Managed]].
+   * Converts this ZIO to [[zio.Managed]]. This ZIO and the provided release action
```
Comment (mschuwalow) [inline core/shared/src/main/scala/zio/ZIO.scala]: This is very nice. Makes it far easier to understand interruption semantics without digging through ZManaged / bracket

**Comment by mschuwalow** [comment]
> @mschuwalow
> Interesting that you bring this up :-) I looked into this when we switched over to `Pull`. My thought then is that racing was less efficient than forking off the streams and having them push into a queue. So this is definitely possible and possibly simpler; worth exploring if you have the time!

Cool. I'll play around with it and open a pr so we can see the differences. btw do we have any relevant benchmarks for this / do you have ideas what to use?

**Comment by iravid** [comment]
No, I don't have anything smart to say about that. If you're using`flatMapPar` you're probably doing some sort of IO, so pure throughput is not very interesting. Allocation rate could be interesting.

**Comment by iravid** [comment]
Any improvement on code complexity would be very welcome in any case. The combinator is super subtle currently.

File: core/shared/src/main/scala/zio/ZManaged.scala [hunk 2]
```diff
@@ -987,12 +1000,20 @@ object ZManaged {
...
    */
   final def fromEffect[R, E, A](fa: ZIO[R, E, A]): ZManaged[R, E, A] =
     ZManaged(IO.succeed(Reservation(fa, _ => IO.unit)))
 
+  /**
+   * Lifts a ZIO[R, E, A] into ZManaged[R, E, A] with no release action. The
+   * effect will be performed uninterruptibly. You usually want the [[ZManaged.fromEffect]]
+   * variant.
+   */
+  final def fromEffectUninterruptible[R, E, A](fa: ZIO[R, E, A]): ZManaged[R, E, A] =
```
Comment (vasilmkd) [inline core/shared/src/main/scala/zio/ZManaged.scala]: This reminds me of `Semaphore.withPermitManaged`. Cool.

File: streams-tests/jvm/src/test/scala/zio/stream/StreamSpec.scala [hunk 3]
```diff
@@ -1736,31 +1755,12 @@ class StreamSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRu
       .runCollect must_=== List()
   }
 
-  private def throttleEnforceShortCircuits = {
```
Comment (vasilmkd) [inline streams-tests/jvm/src/test/scala/zio/stream/StreamSpec.scala]: Good that these are deleted.

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 4]
```diff
@@ -932,29 +932,19 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
   final def flatMap[R1 <: R, E1 >: E, B](f0: A => ZStream[R1, E1, B]): ZStream[R1, E1, B] =
```
Comment (vasilmkd) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: Finally we get to reason about `flatMap` on a higher level. Love it.

File: streams-tests/jvm/src/test/scala/zio/stream/StreamSpec.scala (hunk 3 above)
Comment (iravid) [reply streams-tests/jvm/src/test/scala/zio/stream/StreamSpec.scala]: Ah I had to get rid of something because of the `UTF8 string too long` error.

**Comment by iravid** [comment]
Thanks for the review @mschuwalow and @vasilmkd!

File: streams-tests/jvm/src/test/scala/zio/stream/StreamSpec.scala (hunk 3 above)
Comment (vasilmkd) [reply streams-tests/jvm/src/test/scala/zio/stream/StreamSpec.scala]: Finally, I thought it was just me and I was going crazy.

**Pull request merged**